### PR TITLE
Add EnvironmentLogger Callback

### PR DESCRIPTION
**Pull request by josejg**
This is a new callback to simplify logging environment metadata for reproducibility purposes:

- Git Commits for packages under workspace_dir, useful for mcli integrations
- Package versions for core dependencies, e.g. transformers
- Nvidia versions, including cuda, cuddn and nccl
- Docker image and tag
- Distributed info
- System info 
- Gpu info

Main change is that GitPython now becomes a dependency. It is already a composer[dev] dependency. I was not sure if it was worth gating under a additional dependency

### CODE REVIEW

**Comment by josejg** [comment]
I tested it with both MLflow and Wandb, where it produced the following 

## MLFlow
image

## Wandb
image

**Review by mvpatel2000 — COMMENTED** [review]
Should this be included in the state data and logged? @eracah

**Comment by josejg** [comment]
Pushed an update fixing the failing checks. Let me know if it's good to go or if there's anything that needs to be changed

File: llmfoundry/callbacks/envlogger.py
Comment (b-chu) [inline llmfoundry/callbacks/envlogger.py:1]: Rename file to env_logging_callback

File: llmfoundry/callbacks/envlogger.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+import git
+import pkg_resources
+import psutil
+import torch
+from composer.core import Callback, State
+from composer.loggers import Logger
+
+
+class EnvironmentLoggerCallback(Callback):
```
Comment (b-chu) [inline llmfoundry/callbacks/envlogger.py:17]: Rename to EnvironmentLoggingCallback

File: llmfoundry/callbacks/envlogger.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        log_docker (bool): Whether to log Docker container information. Defaults to True.
+        log_system (bool): Whether to log system information. Defaults to False.
+        log_gpu (bool): Whether to log GPU information. Defaults to False.
+        log_distributed (bool): Whether to log distributed training information. Defaults to False.
+        packages_to_log (list[str]): A list of package names to log versions for. Defaults to None.
+
+    The collected information is logged as hyperparameters at the start of model fitting.
+    """
+
+    PACKAGES_TO_LOG = [
```
Comment (b-chu) [inline llmfoundry/callbacks/envlogger.py:38]: Make private module level constant

File: setup.py [hunk 3]
```diff
@@ -75,6 +75,7 @@
     'tenacity>=8.2.3,<9',
     'catalogue>=2,<3',
     'typer<1',
+    'GitPython==3.1.43',
```
Comment (b-chu) [inline setup.py:78]: Can you make this a range?

File: llmfoundry/callbacks/envlogger.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            }
+
+        return {'available': False}
+
+    def _get_distributed_info(self) -> dict[str, Any]:
+        return {
+            'world_size': int(os.environ.get('WORLD_SIZE', 1)),
+            'local_world_size': int(os.environ.get('LOCAL_WORLD_SIZE', 1)),
+            'rank': int(os.environ.get('RANK', 0)),
+            'local_rank': int(os.environ.get('LOCAL_RANK', 0)),
```
Comment (b-chu) [inline llmfoundry/callbacks/envlogger.py:134]: Get from composer instead?

File: llmfoundry/callbacks/envlogger.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        }
+
+    def _get_docker_info(self) -> dict[str, Any]:
+        from mcli import sdk
```
Comment (b-chu) [inline llmfoundry/callbacks/envlogger.py:138]: Is this not a new dependency?

File: llmfoundry/callbacks/envlogger.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            self.env_data['system_info'] = self._get_system_info()
+
+        if self.log_gpu:
+            self.env_data['gpu_info'] = self._get_gpu_info()
+
+        if self.log_distributed:
+            self.env_data['distributed_info'] = self._get_distributed_info()
+
+        # Log the collected data
+        logger.log_hyperparameters({'environment_data': self.env_data})
```
Comment (b-chu) [inline llmfoundry/callbacks/envlogger.py:177]: Does this get logged to mosaciml logger?

**Review by b-chu — COMMENTED** [review]
Some minor comments and questions. Agreed with Mihir, it seems useful to incorporate this as state dict metadata instead of relying on loggers.

File: setup.py (hunk 3 above)
Comment (josejg) [reply setup.py:78]: I followed the default from composer https://github.com/mosaicml/composer/blob/main/setup.py#L140. Do we want to change it there too?

File: llmfoundry/callbacks/envlogger.py (hunk 6 above)
Comment (josejg) [reply llmfoundry/callbacks/envlogger.py:177]: Yup, https://github.com/mosaicml/composer/blob/main/composer/loggers/mosaicml_logger.py#L94

File: llmfoundry/callbacks/envlogger.py
Comment (josejg) [reply llmfoundry/callbacks/envlogger.py:1]: done!

File: llmfoundry/callbacks/envlogger.py (hunk 1 above)
Comment (josejg) [reply llmfoundry/callbacks/envlogger.py]: done!

File: llmfoundry/callbacks/envlogger.py (hunk 2 above)
Comment (josejg) [reply llmfoundry/callbacks/envlogger.py]: done!

File: llmfoundry/callbacks/envlogger.py (hunk 4 above)
Comment (josejg) [reply llmfoundry/callbacks/envlogger.py]: done!

File: llmfoundry/callbacks/envlogger.py (hunk 5 above)
Comment (josejg) [reply llmfoundry/callbacks/envlogger.py]: It's not (https://github.com/mosaicml/composer/blob/main/setup.py#L93) made a top level import

File: llmfoundry/callbacks/env_logging_callback.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    def _get_distributed_info(self) -> dict[str, Any]:
+        return {
+            'world_size': dist.get_world_size(),
+            'local_world_size': dist.get_local_world_size(),
+            'rank': dist.get_global_rank(),
+            'local_rank': dist.get_local_rank(),
+        }
+
+    def _get_docker_info(self) -> dict[str, Any]:
+        run = sdk.get_run(os.environ['RUN_NAME'])
```
Comment (dakinggg) [inline llmfoundry/callbacks/env_logging_callback.py:140]: probably want to try except this one so it can be run off platform?

**Review by dakinggg — APPROVED** [review]
lets mark this as experimental for now. we may incorporate it into composer directly.

File: llmfoundry/callbacks/env_logging_callback.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        self.log_nvidia = log_nvidia
+        self.log_docker = log_docker
+        self.log_system = log_system
+        self.log_gpu = log_gpu
+        self.log_distributed = log_distributed
+        self.env_data: dict[str, Any] = {}
+        self.packages_to_log = packages_to_log or _PACKAGES_TO_LOG
+
+    def _get_git_info(self, repo_path: str) -> dict[str, str]:
+        repo = git.Repo(repo_path)
```
[the remark this answers is not in the document]
Comment (josejg) [reply llmfoundry/callbacks/env_logging_callback.py:79]: sounds good

File: llmfoundry/callbacks/env_logging_callback.py (hunk 7 above)
Comment (josejg) [reply llmfoundry/callbacks/env_logging_callback.py:141]: ok, but I'd rather check for RUN_NAME

File: setup.py (hunk 3 above)
Comment (josejg) [reply setup.py:76]: Will stick with Composer's choice not to have conflicting deps

**Pull request merged**